packages/notifications/push/amplify_push_notifications/android/src/main/kotlin/com/amazonaws/amplify/amplify_push_notifications/AmplifyPushNotificationsPlugin.kt

```diff
@@ -3,110 +3,263 @@
+                        )
+                        apply()
+                    }
+                }
+                result.success(false)
             }
-            // Get new FCM registration token
-            val token = task.result
-            val hashMap: HashMap<String, Any?> = HashMap()
-            hashMap["token"] = token
-            StreamHandlers.tokenReceived.send(
-                hashMap
+        }
+    }
+
+    override fun getBadgeCount(): Long {
+        throw NotImplementedError("Get badge count is not supported on Android")
+    }
+
+    override fun registerCallbackFunction(
+        callbackHandle: Long,
+        callbackType: PushNotificationsHostApiBindings.CallbackType
+    ) {
+        when(callbackType){
+            PushNotificationsHostApiBindings.CallbackType.DISPATCHER -> registerCallbackToCache(
+                applicationContext,
+                callbackHandle,
+                PushNotificationPluginConstants.CALLBACK_DISPATCHER_HANDLE_KEY
             )
-        })
+            PushNotificationsHostApiBindings.CallbackType.EXTERNAL_FUNCTION -> registerCallbackToCache(
+                applicationContext,
+                callbackHandle,
+                PushNotificationPluginConstants.BG_EXTERNAL_CALLBACK_HANDLE_KEY
+            )
+        }
+    }
 
+    override fun setBadgeCount(withBadgeCount: Long) {
+        throw NotImplementedError("Set badge count is not supported on Android")
     }
 
 
-    // TODO: update this function to be more robust
-    override fun onNewIntent(intent: Intent): Boolean {
-        Log.d(TAG, "onNewIntent in push plugin $intent")
+    private fun registerCallbackToCache(
+        context: Context,
+        callbackHandle: Long,
+        callbackKey: String,
+    ) {
+            Log.i(TAG, "Registering callback function with key $callbackKey")
+            context.getSharedPreferences(
+                PushNotificationPluginConstants.SHARED_PREFERENCES_KEY,
+                Context.MODE_PRIVATE
+            ).edit()
+                .putLong(callbackKey, callbackHandle).apply()
 
-        // TODO: "Decide if we need to add a flag for notification open"
-//        val appOpenedThroughTap = intent.getBooleanExtra("appOpenedThroughTap", false)
+    }
 
-        intent.extras?.let {
-            val remoteMessage = RemoteMessage(it)
-            val notificationHashMap = remoteMessage.asPayload().asChannelMap()
-
-            Log.d(TAG, "Send onNotificationOpened message received event: $notificationHashMap")
-//            PushNotificationEventsStreamHandler.sendEvent(
-//                PushNotificationsEvent(
-//                    NativeEvent.NOTIFICATION_OPENED,
-//                    notificationHashMap
-//                )
-//            )
+    private fun refreshToken() {
+        FirebaseMessaging.getInstance().token.addOnCompleteListener { task ->
+            if (!task.isSuccessful) {
+                task.exception?.let { StreamHandlers.tokenReceived.sendError(it) }
```

Comment:
It cannot be null at this particular call site since we are checking for `!task.isSuccessful`. `task.exception` is null only when the task was not completed fully or has completed successfully according to the [docs](https://developers.google.com/android/reference/com/google/android/gms/tasks/Task#public-abstract-exception-getexception). 

In any case, I changed it to log an error if it is null. 

```dart
if (!task.isSuccessful) {
    if (task.exception == null) {
        Log.e(TAG, "UnknownError: fetching device token.")
    }else{
        StreamHandlers.tokenReceived?.sendError(task.exception!!)
    }
    return@addOnCompleteListener
}
```